### PR TITLE
Add tests for -H and -x input arguments

### PR DESCRIPTION
**Pull request by amrutagokhale**
I added an initial commit for positive tests for 'connector add -H' and 'connector remove -H'. I am nowhere near adding all the tests. But it would be nice if I get some feedback so as to know if I am on the right track.

I will add more commits as I write more tests.

### CODE REVIEW

File: tests/product/test_connectors.py [hunk 1]
```diff
@@ -61,6 +61,96 @@ def test_basic_connector_add_remove(self):
...
+                           self.STANDALONE_PRESTO_CLUSTER)
+        self.run_prestoadmin('server start')
+        for host in self.cluster.all_hosts():
+            self.assert_has_default_connector(host)
+
+        self._assert_connectors_loaded([['system'], ['tpch']])
+
+        self.run_prestoadmin('connector remove tpch -H %(master)s')
+        self.run_prestoadmin('server restart')
+        self.assert_path_removed(self.cluster.master,
```
Comment (rschlussel) [inline tests/product/test_connectors.py]: I see what you mean about -H not making sense.  It's really weird behavior to delete the presto-admin connector config file when you are only removing the connector on one node.  But I still stand by what I said before.

File: tests/product/test_connectors.py [hunk 2]
```diff
@@ -61,6 +61,96 @@ def test_basic_connector_add_remove(self):
...
+            'connector.name=tpch',
+            os.path.join(constants.CONNECTORS_DIR, 'tpch.properties'),
+            self.cluster.master
+        )
+        self.run_prestoadmin('connector add tpch -H %(slave1)s')
+        self.run_prestoadmin('server restart')
+        self.assert_has_connector(self.cluster.slaves[0], 'tpch')
+        self._assert_connectors_loaded([['system'], ['tpch']])
+
+    def test_basic_connector_add_remove_workers_using_dash_h(self):
```
Comment (rschlussel) [inline tests/product/test_connectors.py]: I would make this name more distinct from the one above that tests one worker.  The extra s is easy to miss.  Also I think the word 'basic' isn't particularly informative.  I know lots of tests are named that way, but for brevity's sake I'd get rid of it in these tests.  It's anyway not basic anymore once you are using -H.

File: tests/product/test_connectors.py (hunk 1 above)
Comment (amrutagokhale) [inline tests/product/test_connectors.py]: Yes, I believe that in the long run, we should think about whether -H makes sense for each of the commands, and if it doesn't, we should error out. I will add tests for now anyways.

**Comment by rschlussel** [comment]
these look good.  definitely the right track!

File: tests/product/test_connectors.py [hunk 3]
```diff
@@ -61,6 +61,96 @@ def test_basic_connector_add_remove(self):
...
+            'connector.name=tpch',
+            os.path.join(constants.CONNECTORS_DIR, 'tpch.properties'),
+            self.cluster.master
+        )
+        self.run_prestoadmin('connector add tpch -H %(slave1)s')
+        self.run_prestoadmin('server restart')
+        self.assert_has_default_connector(self.cluster.slaves[0])
+        self._assert_connectors_loaded([['system'], ['tpch']])
+
+    def test_basic_connector_add_remove_workers_using_dash_h(self):
```
Comment (rschlussel) [inline tests/product/test_connectors.py]: I would make this name more distinct from the one above that tests one worker. The extra s is easy to miss. Also I think the word 'basic' isn't particularly informative. I know lots of tests are named that way, but for brevity's sake I'd get rid of it in these tests. It's anyway not basic anymore once you are using -H.

File: tests/product/test_connectors.py (hunk 2 above)
Comment (amrutagokhale) [inline tests/product/test_connectors.py]: done

File: tests/product/test_connectors.py [hunk 4]
```diff
@@ -61,6 +61,96 @@ def test_basic_connector_add_remove(self):
...
+            'connector.name=tpch',
+            os.path.join(constants.CONNECTORS_DIR, 'tpch.properties'),
+            self.cluster.master
+        )
+        self.run_prestoadmin('connector add tpch -H %(slave1)s')
+        self.run_prestoadmin('server restart')
+        self.assert_has_default_connector(self.cluster.slaves[0])
+        self._assert_connectors_loaded([['system'], ['tpch']])
+
+    def test_connector_add_remove_two_workers_using_dash_h(self):
```
Comment (kokosing) [inline tests/product/test_connectors.py]: I don't know why you have 3 tests of the same functionality. Does it matter what node you add or remove? I think the last tests is enough. I just would mix slave with master, and one test could be enough. Less tests, faster it is executed, less code needs to be maintained.

What do you think?

Comment (ebd2) [inline tests/product/test_connectors.py]: Agree wholeheartedly. The product tests are a very expensive way to find bugs. This tests the behavior of the command line parsing more than the actual product tests.

Comment (amrutagokhale) [inline tests/product/test_connectors.py]: Good point. I agree that the last test should be sufficient. The latest commit retains only the last test.

**Comment by kokosing** [comment]
looks good

**Comment by amrutagokhale** [comment]
@kokosing @ebd2 Added three more commits. Can you review?

File: tests/product/test_collect.py [hunk 5]
```diff
@@ -64,6 +64,14 @@ def test_collect_logs_basic(self):
         admin_log = path.join(downloaded_logs_location, PRESTOADMIN_LOG_NAME)
         self.assert_path_exists(self.cluster.master, admin_log)
 
+    def test_collect_logs_dash_h_invalid_host(self):
+        self.setup_cluster(NoHadoopBareImageProvider(),
+                           self.STANDALONE_PRESTO_CLUSTER)
+        expected = 'Hosts defined in --hosts/-H must be present in /etc/opt/prestoadmin/config.json'
```
Comment (ebd2) [inline tests/product/test_collect.py]: This looks too long. Have you run `make lint`?

File: tests/product/test_collect.py [hunk 6]
```diff
@@ -131,6 +139,28 @@ def test_basic_system_info(self, coordinator=None, hosts=None):
...
+        self.test_basic_system_info(actual, self.cluster.internal_master, [self.cluster.master, self.cluster.slaves[0]])
+
+    def test_collect_system_info_dash_x_two_workers(self):
+        self.setup_cluster(NoHadoopBareImageProvider(),
+                           self.STANDALONE_PRESTO_CLUSTER)
+        self.run_prestoadmin('server start')
+        actual = self.run_prestoadmin('collect system_info -x %(slave2)s,%(slave3)s')
+        self.test_basic_system_info(actual, self.cluster.internal_master, [self.cluster.master, self.cluster.slaves[0]])
+
+    def test_collect_system_info_dash_h_invalid_host(self):
```
Comment (ebd2) [inline tests/product/test_collect.py]: Validating the command line arguments should be tested in a unit test; this shouldn't vary from command to command, and it *definitely* shouldn't rely on having a cluster up.

File: tests/product/test_configuration.py [hunk 7]
```diff
@@ -107,6 +115,70 @@ def test_configuration_deploy_show(self):
         self.assert_node_config(self.cluster.master,
                                 self.default_node_properties_)
 
+    def test_configuration_deploy_using_dash_h_coord_worker(self):
+        self.upload_topology()
+
+        self.deploy_and_assert_default_config()
+
+        dummy_prop1, dummy_prop2 = self.write_dummy_config_file()
```
Comment (ebd2) [inline tests/product/test_configuration.py]: Ooh, I like this! Nice touch!

**Comment by ebd2** [comment]
Apart from getting rid of the invalid host cases and writing a single unit test, this looks good.

File: tests/product/test_configuration.py [hunk 8]
```diff
@@ -58,15 +55,26 @@ def test_configuration_deploy_show(self):
 
         self.assertEqualIgnoringOrder(output, expected)
 
+    def write_dummy_config_file(self):
```
Comment (kokosing) [inline tests/product/test_configuration.py]: make this private by prefixing with `__`

**Comment by kokosing** [comment]
lgtm

File: tests/product/test_collect.py (hunk 6 above)
Comment (kokosing) [inline tests/product/test_collect.py]: +1

File: tests/product/test_connectors.py [hunk 9]
```diff
@@ -90,6 +90,22 @@ def test_connector_add_remove_coord_worker_using_dash_h(self):
         self.assert_has_default_connector(self.cluster.master)
         self.assert_has_default_connector(self.cluster.slaves[1])
 
+    def test_connector_add_invalid_host_using_dash_h(self):
```
Comment (kokosing) [inline tests/product/test_connectors.py]: these two tests looks like they could be written at unit test level as they are just parsing arguments

File: tests/product/test_configuration.py (hunk 8 above)
Comment (amrutagokhale) [inline tests/product/test_configuration.py]: done

File: tests/product/test_collect.py (hunk 6 above)
Comment (amrutagokhale) [inline tests/product/test_collect.py]: That's a good suggestion. I am now adding a single unit test for one sample command `server uninstall -H <invalid host name>` and I am going to remove all the product tests for `-H <invalid host name>`. There is no particular reason for choosing `server uninstall` command. I had to choose one command to which I can provide `-H` option. `server uninstall` already had an existing unit test that checks for invalid host name using --hosts: `server uninstall --hosts <invalid host name>`. So I added the test for the shorter `-H` option for the same command.

File: tests/product/test_collect.py (hunk 5 above)
Comment (amrutagokhale) [inline tests/product/test_collect.py]: This is being removed and replaced by a single unit test.

File: tests/product/test_connectors.py (hunk 9 above)
Comment (amrutagokhale) [inline tests/product/test_connectors.py]: Yes, I have removed these as well. Now, there exists a single unit test that checks for `-H <invalid host name>`

**Comment by amrutagokhale** [comment]
updated

**Comment by kokosing** [comment]
thanks for applying comments, LGTM :)

**Comment by kokosing** [comment]
Fix travis build, it looks like `lint` found some issues.

**Comment by amrutagokhale** [comment]
Merged https://github.com/prestodb/presto-admin/pull/148

**Pull request merged**